File: src/core/notify.ts
type Notifier = (title: string, message: string) => void;

function osascriptNotifier(title: string, message: string) {
  const esc = (s: string) => s.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
  Bun.spawnSync(["osascript", "-e", `display notification "${esc(message)}" with title "${esc(title)}"`]);
}

let notifier: Notifier = osascriptNotifier;

/** The menu bar app swaps in its native notifications so they show under the app's name. */
export function setNotifier(fn: Notifier) {
  notifier = fn;
}

export function notify(title: string, message: string) {
  try {
    notifier(title, message);
  } catch {
    // notifications are best-effort
  }
}

File: src/go/core/notify.go
package core

import (
	"fmt"
	"os/exec"
	"strings"
)

// Notifier shows a notification. The menu bar app swaps in its native notifications
// so they show under the app's name.
var Notifier = func(title, message string) {
	esc := strings.NewReplacer(`\`, `\\`, `"`, `\"`)
	_ = exec.Command("osascript", "-e",
		fmt.Sprintf(`display notification "%s" with title "%s"`, esc.Replace(message), esc.Replace(title))).Run()
}

// Notify is best-effort.
func Notify(title, message string) {
	defer func() { _ = recover() }()
	Notifier(title, message)
}
